### PR TITLE
Add plugincontainer module

### PR DESCRIPTION
**Pull request by tomhjp**
Follows up on the new interfaces added in https://github.com/hashicorp/go-plugin/pull/270.

Apologies for the size, although a lot of it is generated/license code. Happy to take feedback on splitting this up into multiple PRs if preferred. Maybe the example could be added separately for instance, but it does make for quite a nice end to end integration test of the feature.

### CODE REVIEW

**Comment by tomhjp** [comment]
I think this is in a decent state to review now. I'm planning to tag go-plugin and pull that tag in once there's rough consensus on this PR, just in case there are any last minute API changes required in go-plugin.

**Comment by tomhjp** [comment]
Note: Looks like the CI gives a green tick even if the build and tests fail: https://github.com/hashicorp/go-secure-stdlib/actions/runs/5956014581/job/16155887975. I'll look into fixing this in a separate PR tomorrow.

File: plugincontainer/container_runner.go [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+	}
+	if cmd.Args != nil {
+		containerConfig.Cmd = cmd.Args
+		containerConfig.ArgsEscaped = true
+	}
+
+	// Host config.
+	// TODO: Can we safely we drop some default capabilities?
+	hostConfig := &container.HostConfig{
+		AutoRemove:    true,                      // Plugin containers are ephemeral.
```
Comment (fairclothjm) [inline plugincontainer/container_runner.go:84]: Is this behavior possible due to Vault's lazy-loading? i.e. if the container is not running, Vault will restart it when the plugin receives a request.

File: plugincontainer/container_runner_test.go [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+	}
+
+	// Request the plugin
+	raw, err := rpcClient.Dispense("counter")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// We should have a Counter store running inside a container now! This feels
+	// like a normal interface implementation but is in fact over an RPC connection.
```
Comment (fairclothjm) [inline plugincontainer/container_runner_test.go:163]: 🎉

**Review by fairclothjm — APPROVED** [review]
LGTM!

File: plugincontainer/container_runner.go (hunk 1 above)
Comment (tomhjp) [reply plugincontainer/container_runner.go:84]: I'm not sure I quite understand the question. The reasoning for this setting is to reduce management overhead - we _could_ keep restarting the same container, but then we need to be more careful about tracking when we can delete it. Whereas if we just always delete containers after the plugin process has exited, then it mirrors the lifecycle of a plain plugin process more closely.

Comment (tomhjp) [reply plugincontainer/container_runner.go:84]: It's equivalent to the `--rm` setting for `docker run` btw

Comment (fairclothjm) [reply plugincontainer/container_runner.go:84]: That makes sense, thanks!

**Comment by tomhjp** [comment]
Sorry for the commit noise, I accidentally pushed a merge with my CI branch, and chose to revert the merge commit rather than rebase so that it's easier to see changes since last review.

**Comment by tomhjp** [comment]
I think this is relatively stable now, but I'm yet to finish the Vault changes leveraging this PR, so I might discover some more requirements. There also seems to be a problem with the tests failing in CI, but they work on my machine, so that may take a bit of time to unpick. ([Failing CI link with #85 merged in](https://github.com/hashicorp/go-secure-stdlib/actions/runs/5969616136))

**Review by jefferai — APPROVED** [review]
Approving from a repo coordinator perspective, thanks for keeping me in the loop 👍

**Review by briankassouf — APPROVED** [review]
This looks fantastic!

**Comment by tomhjp** [comment]
Thanks all!

**Pull request merged**